### PR TITLE
Implement Random.cpp for MSVC

### PR DESCRIPTION
**Pull request by Orvid**
This uses `<random>` to implement it, because there is no `/dev/urandom` on Windows.

### CODE REVIEW

**Comment by sgolemon** [comment]
From @yfeldblum ...

This looks deterministic. See: http://www.cplusplus.com/reference/random/mersenne_twister_engine/mersenne_twister_engine/.

That is not OK for readRandomDevice, even on Windows. Windows has CryptGenRandom, though - perhaps that's something that can be used here.

File: folly/Random.cpp [hunk 1]
```diff
@@ -26,16 +26,33 @@
...
+#include <wincrypt.h>
+#endif
+
 namespace folly {
 
 namespace {
 
 void readRandomDevice(void* data, size_t size) {
+#ifdef _MSC_VER
+  static bool initialized = false;
```
Comment (sgolemon) [inline folly/Random.cpp]: (From @yfeldblum) This is not threadsafe. Two threads that simultaneously call readRandomDevice can overwrite each other's data. Using std::once_flag should help.

```
#include <mutex>

static std::once_flag flag;
static HCRYPTPROV provider;
std::call_once(flag, [&] }
  PCHECK(CryptAcquireContext(&provider, nullptr, nullptr, PROV_RSA_FULL, 0));
});
CHECK(size <= std::numeric_limits<DWORD>::max()); // not PCHECK
PCHECK(CryptGenRandom(provider, (DWORD)size, (BYTE*)data);
```

The other possibility is to use ThreadLocal.

Comment (Orvid) [inline folly/Random.cpp]: That was actually deliberate, and is a result of my own personal form of insanity. As the impact of a second call to `CryptoAcquireContext` is minimal on resource usage, basically just incrementing a reference count, and as `cryptoProv` will be entirely written from a single instruction, there is no possibility of getting part of one thread's info and part of another's. It's perfectly fine for a second thread to be trying to initialize the provider at the same time, that's why it's setting initialized after the provider has been acquired.

If it's preferred to use a mutex though, I can certainly do that?

Comment (yfeldblum) [inline folly/Random.cpp]: Is this behavior of `CryptAcquireContext` specified or documented somewhere?

Comment (Orvid) [inline folly/Random.cpp]: I believe my thought process on that might have come from the way `CryptReleaseContext` is documented. Either way, I've switch this to a proper mutex instead.

**Pull request closed without merging**